openwisp_controller/config/admin.py

```diff
@@ -886,6 +892,34 @@ def get_urls(self):
     def schema_view(self, request):
         return JsonResponse(app_settings.DEVICE_GROUP_SCHEMA)
 
+    def add_view(self, request, form_url='', extra_context=None):
+        extra_context = extra_context or {}
+        extra_context.update(self.get_extra_context())
+        return super().add_view(request, form_url, extra_context)
+
+    def change_view(self, request, object_id, form_url='', extra_context=None):
+        extra_context = self.get_extra_context(object_id)
+        return super().change_view(request, object_id, form_url, extra_context)
+
+    def save_related(self, request, form, formsets, change):
+        old_templates = list(form.instance.templates.values_list('pk', flat=True))
+        super().save_related(request, form, formsets, change)
+        templates = list(form.instance.templates.values_list('pk', flat=True))
+        DeviceGroup.templates_changed(
+            form.instance, old_templates=old_templates, templates=templates
+        )
```

Comment:
I tried `m2m_changed` signal also, but in that signal it is hard to get old templates, so I have to do this work around to get old templates before `form.save_m2m` and new templates after `save_m2m`

In `m2m_changed` signal the only possibility to make it work was by using `pre_clear` and `post_save` action i.e. clearing all applied templates first and then reapplying it.